### PR TITLE
Homework 28.1

### PR DESCRIPTION
**Pull request by vremere-a**

### CODE REVIEW

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        String url = req.getServletPath();
+        if (url.equals("/login") || url.equals("/registration")) {
+            chain.doFilter(req, resp);
+            return;
+        }
+
+        Long userId = (Long) req.getSession().getAttribute(USER_ID);
+        if (userId == null || userService.getById(userId) == null) {
+            resp.sendRedirect("/login");
```
Comment (fyarkiy) [inline src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]: I think you can do here 
resp.sendRedirect(req.getContextPath() + "/login") - Bohdan just missed it on video

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            return;
+        }
+        chain.doFilter(req, resp);
+    }
+
+    @Override
+    public void destroy() {
+
```
Comment (fyarkiy) [inline src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]: no need in this empty line

**Review by fyarkiy — APPROVED** [review]

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java (hunk 1 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]: done

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java (hunk 2 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]: done

File: src/main/java/com/internet/shop/controllers/user/LoginController.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String login = req.getParameter("login");
+        String password = req.getParameter("pwd");
+        try {
+            User user = authService.login(login, password);
+            HttpSession session = req.getSession();
+            session.setAttribute("user_id",user.getId());
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/controllers/user/LoginController.java]: let's make it constant as in filter

File: src/main/java/com/internet/shop/dao/implementations/UserDaoImpl.java [hunk 4]
```diff
@@ -40,4 +40,10 @@ public boolean deleteById(Long id) {
     public List<User> getAll() {
         return Storage.users;
     }
+
+    @Override
+    public Optional<User> findByLogin(String login) {
+        return Storage.users.stream()
+                .filter(s -> s.getLogin().equals(login)).findFirst();
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/implementations/UserDaoImpl.java]:
```suggestion
        return Storage.users.stream()
                .filter(u -> u.getLogin().equals(login))
                .findFirst();
```

File: src/main/java/com/internet/shop/security/implementations/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Override
+    public User login(String login, String password) throws AuthenticationException {
+        User userFromDB = userService.findByLogin(login).orElseThrow(() ->
+                new AuthenticationException("Incorrect username or password"));
+
+        if (userFromDB.getPassword().equals(password)) {
+            return userFromDB;
+        }
+        throw new AuthenticationException("Incorrect username or password");
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/security/implementations/AuthenticationServiceImpl.java]: either make all checks in single if(and throw exception only once)  or make exception message - constant

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+@WebFilter("/*")
+public class AuthenticationFilter implements Filter {
+    private static final String USER_ID = "user_id";
+    private static final Injector injector = Injector.getInstance("com.internet.shop");
+    private final UserService userService = (UserService) injector.getInstance(UserService.class);
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]:
```suggestion
```

**Review by dmytro-hryhoruk — COMMENTED** [review]
push changes in web.xml too

File: src/main/java/com/internet/shop/controllers/user/LoginController.java (hunk 3 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/controllers/user/LoginController.java]: done

File: src/main/java/com/internet/shop/dao/implementations/UserDaoImpl.java (hunk 4 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/dao/implementations/UserDaoImpl.java]: done

File: src/main/java/com/internet/shop/security/implementations/AuthenticationServiceImpl.java (hunk 5 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/security/implementations/AuthenticationServiceImpl.java]: done

File: src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java (hunk 6 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/web/filters/AuthenticationFilter.java]: done

**Review by dmytro-hryhoruk — COMMENTED** [review]
You have created a new controller and filter, which you should have added to web.xml(mappings for them). But I don't see any changes in web.xml file in this PR

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**